src/data/set/finite.lean

```diff
@@ -20,529 +47,733 @@ variables {α : Type u} {β : Type v} {ι : Sort w} {γ : Type x}
 
 namespace set
 
-/-- A set is finite if the subtype is a fintype, i.e. there is a
-  list that enumerates its members. -/
+/-- A set is finite if there is a `finset` with the same elements.
+This is represented as there being a `fintype` instance for the set
+coerced to a type. -/
 inductive finite (s : set α) : Prop
 | intro : fintype s → finite
 
-lemma finite_def {s : set α} : finite s ↔ nonempty (fintype s) := ⟨λ ⟨h⟩, ⟨h⟩, λ ⟨h⟩, ⟨h⟩⟩
+/-- Constructor for `set.finite` with the `fintype` as an instance argument. -/
```

Comment:
Can you make `intro` be this with something like
```
inductive finite (s : set α) : Prop
| of_fintype [fintype s] : finite
```
Or is that annoying?

Reply:
I'd found writing `finite_def` awkward that way, so I figured I'd leave it as Yury wrote it.